docs/design/uplink-contexts.md

```diff
@@ -0,0 +1,103 @@
+# Title: Uplink Contexts
+
+## Abstract
+
+This design document proposes the data we need to store inside of a share in order to access files. Currently, the uplink configuration yaml file includes distinct entries for the API key, the satellite url, and the root encryption key. This proposal is to bundle an API key, a satellite url, and a list of revealed paths into a context (much like a kubectl context) and call that bundle a share.
+
+## Background
+
+In order to access objects, one must provide an API key to a given satellite. This API key is tied to a project which allows the satellite to look up the specific bucket being requested and do the things it needs to do. API keys are moving to be macaroons where they can be limited to only allow reads or writes, or expire, etc. Those features can easily be implemented and don't interact much the overall uplink command experience: you have access to do some actions or you don't.
+
+There are complications when it comes to restricting to a whitelist of buckets and/or path prefixes inside of those buckets. For example, if a macaroon is limited to only allow access to paths beneath `sj://bucket/a/b`, then it would be impossible to list the root. Additionally, the clients and servers all operate on _encrypted_ paths, making it impossible for the server to issue the list and filter, because one would have to reveal the root encryption key for the bucket for the client to be able to decrypt the names.
+
+One could attempt to cause restricted macaroons to be "mounted". For example, if a bucket contains `sj://bucket/a/b/c` then the listing of the root would return "c". Unfortunately, this does not compose well with having multiple allowed paths in a macaroon: which "root" do you mean? It also causes confusion because the paths of objects inside of a bucket becomes a local rather than global property, even though the clients and servers must speak with the global view at all times.
+
+Additionally, there's a single global key per project. That's less than ideal for a situation like having distinct prod and staging buckets: you don't want them both using the same encryption keys so that giving access to staging does not leak the encryption keys for prod. It also doesn't allow one to re-key a bucket if a share is leaked. Even if the macaroon inside is revoked, it may be seen as a risk that the key was leaked because they are now depending on the satellite operator for the safety of their data.
```

Comment:
The problem I'm describing is that sj://prod and sj://staging have the same root encryption key, so all their path based keys are the same as well. So if you share sj://staging/foo/bar with someone, that also leaks the key for sj://prod/foo/bar.

Is my understanding wrong? Maybe the bucket name is included in the derivation of the root key? That still has revocation issues, though.

Like, right now, I only need one secret to get to all of the buckets inside of a project. That either gives up cross bucket secrecy, or bucket level revocation of keys.

Reply:
Yeah, bucket name is included in the derivation.

> Like, right now, I only need one secret to get to all of the buckets inside of a project.

I don't know that we can avoid that scenario. It seems congruent to the AWS situation, where you only need the API key for a root account to be able to get at anything stored under that account. The best solution would seem to be "don't give people the root encryption key, then".

Revocation not being able to undo encryption key access is a thing that has bothered me, though. I don't know what solutions we can even offer. Recommend that people re-key their projects and reencrypt+reupload everything in them anytime somebody leaves the project? Add some kind of generational-key feature where everything added before time T uses key K1 but everything after T uses key K2?

Reply:
This proposes to at least have each bucket be independently keyed. Do you think that's worth doing?

Reply:
Some bullet points to help be clear about trade-offs:

1. Sharing the project root key shares into the future as well. Having each bucket keyed does not share anything except what was explicitly shared.
2. It's possible to revoke a bucket key by deleting it, giving it a new key, and uploading it again. To revoke a bucket with project level keys, the entire project would have to undergo a delete and reupload.
3. It's not possible to give a single key that works for all buckets and all time.

Point 1. doesn't require per bucket keys, it just requires not sharing the project key, but the fact that it exists means it's possible to leak. Point 3 could be considered a plus or a minus as far as I can see.

One way to think of it is that currently the bucket is just another path component as far as revocation is concerned, so we're choosing between having a single root key, or multiple root keys. I think multiple root keys is the safer option.

Reply:
> This proposes to at least have each bucket be independently keyed. Do you think that's worth doing?

I think I missed a step. What is proposing that?

Reply:
Oops, you're right in that it doesn't necessarily follow from the data model. I'm explicitly proposing we do that now, though.

Reply:
I think I'm going to reveal my idiocy by not understanding this comment, but oh well it's not a secret anymore

> Some bullet points to help be clear about trade-offs:

I can't seem to parse these into helpful points. Are these meant to be alternate choices, observations about the current state, observations about the code, or proposals?

> 1. Sharing the project root key shares into the future as well. Having each bucket keyed does not share anything except what was explicitly shared.

What does it mean to share into the future? What does it mean to have each bucket keyed? Why would keying a bucket cause something to be shared?

> It's possible to revoke a bucket key by deleting it, giving it a new key, and uploading it again. To revoke a bucket with project level keys, the entire project would have to undergo a delete and reupload.

Not sure we can really call that revoking, but sure, that would be a way to replace a key. I don't think we can treat it as a viable option in most cases.

> It's not possible to give a single key that works for all buckets and all time.

Why is that a thing anyone would expect to have? Are we still talking about encryption keys?

Reply:
It's also possible I'm an idiot so here goes:

In all of the following, I'm assuming a threat model where if someone has access to the encryption key, they have access to the data. In other words, the attacker is in control of the satellite somehow (either a 3rd party or the satellite operator themselves).

> What does it mean to share into the future? What does it mean to have each bucket keyed? Why would keying a bucket cause something to be shared?

By "shared into the future", I mean that if you have a project with some collection of buckets {a, b, c}, and you gain access to the project root key, you also gain access to the key for any bucket created in the future, not just {a, b, c}.

By "having each bucket keyed" I mean that each bucket has a distinct root key to derive all of the path keys.

By saying "having each bucket keyed does not share anything except what was explicitly shared", I mean that there is no value that you can leak that gives the keys to any buckets that have yet to be created.

> Not sure we can really call that revoking, but sure, that would be a way to replace a key. I don't think we can treat it as a viable option in most cases.

I agree, but it's strictly better than having to do the whole project.

> Why is that a thing anyone would expect to have? Are we still talking about encryption keys?

You can currently have it by having the root project key. That key derives all of the keys used for all of the buckets that exist in a project now and into the future. Yes, we're talking about encryption keys.

Reply:
> In all of the following, I'm assuming a threat model where if someone has access to the encryption key, they have access to the data. In other words, the attacker is in control of the satellite somehow (either a 3rd party or the satellite operator themselves). In other other words, the macaroon is useless from a security perspective.

Ok. That is definitely a valid model, but one that is harder to argue for. To compare with AWS again, I don't think anyone expects their data on S3 to remain 100% secure from an attacker that controls the S3 infrastructure. And I don't really expect our project management folks to be too worried about it.

However, there is a closely related threat model where attackers don't have access to the satellite, but do have access to data on many storage nodes (enough to have a non-negligible chance of getting some partial usable target data on their controlled nodes) and do have access to your encryption keys (because of revoked access or the like). In my opinion, we really do need a good answer for this threat model because it's easy to see how feasible such an attack could be. And most of your points here still hold in that model.

> By "having each bucket keyed" I mean that each bucket has a distinct root key to derive all of the path keys.

Possible terminology difficulty: I think they already have distinct root keys, if I am not confusing terms here, but I think you mean "independent" root keys, i.e., root keys not derived from a common source? That is already possible within the existing framework, actually; you can just declare that files in bucket B will now use a new arbitrary encryption key K if you want. To make that work with uplink, I guess you could just make a new scope for bucket B and key K.

> By saying "having each bucket keyed does not share anything except what was explicitly shared", I mean that there is no value that you can leak that gives the keys to any buckets that have yet to be created.

Ok. I don't know that that's too important. It seems to amount to "if you give away the top level secret-est root key and then you want to rescind that access so that the recipients can no longer decrypt any of the things you have stored, even if they get their hands on it, it's going to be a huge pain in the ass and you'll have to re-encrypt everything", which is a property that many other viable systems also have :)

> > Why is that a thing anyone would expect to have? Are we still talking about encryption keys?
>
> You can currently have it by having the root project key. That key derives all of the keys used for all of the buckets that exist in a project now and into the future. Yes, we're talking about encryption keys.

Ohhhh you you're talking about all buckets in that project derived from the same encryption key, not all buckets ever. Ok, makes sense. Definitely a valid concern for the given threat model.

What about this possible solution for this problem of "revocable encryption":

1. Add a field to `BucketShare`: `AlternateKeys map[int16][]byte`
2. Add a field to the pointer protobuf: `int16 alternate_key_id = N;`
3. Add some way to tell `uplink` that I want to use a new encryption key with `alternate_key_id=1` by default, for new objects in the relevant scope
4. Such new objects get `alternate_key_id=1` set in their pointer metadata
5. `uplink` uses the default `EncryptionKey` when decrypting old objects (objects with no alternate_key_id in the metadata) and the new key when decrypting new objects (because `AlternateKeys[1]` has the new key)

An attacker with revoked access could still decrypt objects stored with the old encryption key, but they could already have gotten the contents of those same objects _before_ their access was revoked anyway, so this already isn't a problem we can solve. And the user doesn't have to download, re-encrypt, or re-upload anything.

Reply:
hmm, that's an interesting idea! that does make it so the common case is one doesn't have to configure a passphrase when a new bucket shows up. i do think it would be good to at least have some affordance for that case (either a set of key_id to key mappings so we can persist bucket to key_id mappings or alternatively just a default key your uplink is configured with)?

in general i do think forcing all buckets to use a key derived from a project root encryption key forces some bad experiences (you must know the project root encryption key to make new buckets) but it doesn't sound like we're arguing for that so i'm happy.

Reply:
I don't think that proposal works. The key at the root of the bucket is used to derive a tree of paths. Changing that key would mean changing all of the encrypted paths, so now there's possibly objects that exist that are shadowed. You'd have to list N different encrypted locations if there's N possible alternate keys. That's quite a bit of complexity.

Also, that proposal is orthogonal to if each bucket has an independent root key. One could exist with or without the other.

This design doc does not depend on buckets being independently keyed, so I propose we drop it for now, and continue with a share including the bucket name so that we can support it in the future if we want.

Reply:
:+1: